### PR TITLE
Swaps out the redis implementation for a different shard

### LINKED ISSUE #46: Look at using alternate Redis shard
This shard https://github.com/jgaskins/redis seems to have better support in some areas, and several people have mentioned that it's a little easier to work with. We can look in to what the differences are, and if it would provide any extra stability or improvements.

### PR DESCRIPTION
**Pull request by jwoertink**
Fixes #46 

This PR swaps out the redis implementation to a new shard. With the new shard, the monkey patch is no longer required, and several stability improvements are made. This redis shard includes better connection pooling, and support for things like redis cluster. This also includes a new RedisPinger to ensure redis stays alive.

The main breaking change is that some of the settings have been removed. These settings will need to be removed from your config after upgrading:

```
setting pool_redis_publish : Bool = false
setting redis_pool_size : Int32 = 5
setting redis_pool_timeout : Float64 = 5.0
```

### CODE REVIEW

File: src/cable/server.cr [hunk 1]
```diff
@@ -148,17 +129,23 @@ module Cable
...
+                debug
+              else
+                fiber_channel.send({channel, message})
+                Cable::Logger.debug { "Cable::Server#subscribe channel:#{channel} message:#{message}" }
+              end
             end
           end
+        rescue e : IO::Error
+          # why is redis_subscribe.@socket.closed? here??
         end
```
Comment (jwoertink) [inline src/cable/server.cr:149]: Between each spec we call `Cable.restart`. After the first spec runs, this block here ends up throwing an `IO::Error`. It seems when we try to run `redis_subscribe.subscribe("_internal")`, the `redis_subscribe.@socket` is already `closed?`. It should be a new instance with a new socket, but I'm not sure why it's not... For now, we can ignore it

**Comment by jwoertink** [comment]
The CI failures are because Ameba doesn't support Crystal < 1.4.0. Ameba isn't necessary here. I can either take it back out, or we can just lock Cable up to 1.4.0+ too. (This is what we did with Lucky)

**Comment by jwoertink** [comment]
@jgaskins if you're poking around this, you'll see that even though the specs pass, they throw all those IO::Error https://github.com/cable-cr/cable/runs/7380811910?check_suite_focus=true#step:6:7 if that helps any.

File: src/cable/server.cr (hunk 1 above)
Comment (jgaskins) [reply src/cable/server.cr]: As mentioned in https://github.com/jgaskins/redis/issues/12, I played around with this trying to understand what's going on. I added a bunch of `puts` statements all over the place to see the order of operations.

<details><summary>Changes made to code in this PR and monkeypatches to Redis</summary>

```crystal
# module Cable
  def self.restart
    if current_server = @@server
      puts "Shutting down old server..."
      current_server.shutdown
    end
    puts "Spinning up new server..."
    @@server = Server.new
  end

# class Cable::Server
    # Use the pooled connections
    getter redis_publish = begin
      puts "Connecting publisher"
      Redis::Client.new(URI.parse(Cable.settings.url))
    end
    # Use a single connection
    getter redis_subscribe = begin
      puts "Connecting subscriber"
      Redis::Connection.new(URI.parse(Cable.settings.url))
    end

# in Redis::Connection#initialize
      pp connecting: @uri.to_s
      @writer = Writer.new(IO::MultiWriter.new(STDERR, socket))
```

</details>

There may also be a `Fiber.yield` in there to try to give the spawned fibers a chance to run:

---

Here's the output for an `IO::Error` chosen arbitrarily:

```text
Shutting down old server...
*1
$11
unsubscribe
Spinning up new server...
Connecting publisher
{connecting: "redis://localhost:6379"}
*Unhandled exception in spawn(name: Cable::Server - subscribe): Closed stream (IO::Error)
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io.cr:118:5 in 'check_open'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io/buffered.cr:129:5 in 'write'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io/multi_writer.cr:37:30 in 'write'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io.cr:877:5 in 'write_byte'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/char.cr:897:9 in 'to_s'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io.cr:174:5 in '<<'
  from lib/redis/src/writer.cr:14:7 in 'encode'
  from lib/redis/src/connection.cr:136:7 in 'subscribe'
  from lib/redis/src/connection.cr:334:9 in 'call'
  from lib/redis/src/connection.cr:119:5 in '->'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/fiber.cr:146:11 in 'run'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/fiber.cr:98:34 in '->'
Connecting subscriber
{connecting: "redis://localhost:6379"}
Initializing server...
*2
$9
subscribe
$9
_internal
```

It looks like things are running out of order? That doesn't make any sense, though.

Comment (jwoertink) [reply src/cable/server.cr]: Oh... Is it trying to subscribe to the _internal before the subscribe has even been connected? :thinking:

Comment (jgaskins) [reply src/cable/server.cr]: That's what I thought at first, and I was like "Crystal y u like dis?" and then [this realization hit me](https://github.com/jgaskins/redis/issues/12#issuecomment-1186628415)

File: src/cable/server.cr [hunk 2]
```diff
@@ -24,6 +24,9 @@ module Cable
     # Use a single connection
     getter redis_subscribe = Redis::Connection.new(URI.parse(Cable.settings.url))
     getter fiber_channel = ::Channel({String, String}).new
+    getter pinger : Cable::RedisPinger do
+      Cable::RedisPinger.new(self)
+    end
```
Comment (jgaskins) [inline src/cable/server.cr:29]: 💯

**Comment by jwoertink** [comment]
Good news is the memory leak is fixed.... bad news is there's still some issue where when either the docker container swaps, or redis disconnects, it requires a full reboot to fix. It didn't seem to happen before https://github.com/cable-cr/cable/pull/48/commits/6b11799e8b67adc98c70e0be77a057e096da617e It felt like it was handling autoreconnect just fine, but less than 24 hours after deploying that change I had to reboot the whole machine.

**Comment by jgaskins** [comment]
The folks at EatApp had a Redis disconnect issue related to pub/sub, too, related to message size. Does any of the discussion in https://github.com/jgaskins/redis/issues/7 help?

**Comment by mjeffrey18** [comment]
Hey @jwoertink @jgaskins, I'm going to create another PR with just the error-handling stuff on its own as we added some error rate handling logic which has been super useful.

Although I love the idea of putting @jgaskins Redis shard in here, as we use it also, I guess long term, shall we consider an adapter pattern? There are some differences in the way each Redis shard handles types.
For instance, for hgetall I had to write a converter

```crystal
  def fetch_statuses
    return unless (vals = redis.hgetall(cache_key))
    vals = vals.as(Array(Redis::Value))

    if vals.try(&.empty?)
      nil
    else
      data = {} of String => Int64 | Int32
      vals.each_with_index do |val, idx|
        if idx.even?
          data[val.to_s] = vals[idx + 1].to_s.to_i
        end
      end
      data
    end
  end
```

It was mildly painful but worth it since we have our own version of the cable shard. But for others, upgrading could result in substantial breaking changes. Which maybe just needs to happen anyways to make this more stable.

I was also toying with the idea of having a Postgres PubSub backend a while back when Redis was giving me a massive headache. Rails supports a few other backends. We could have 1 for each of the Redis shards as a v1.
Another thing I was experimenting with was Redis streams, so we don't miss pub/sub messages. 
Could be an interesting approach to allow a buffet.

**Comment by jwoertink** [comment]
I definitely want to add a postgres backend https://github.com/cable-cr/cable/issues/49 but doing a redis backend gets a little tricky. We ran in to this issue with [Mosquito](https://github.com/mosquito-cr/mosquito/pull/94#issuecomment-1187990999). The thing is, we can't include both redis shards in to the same project at the same time. So in order for Cable to have a generic pluggable redis interface, we'd need some sort of mock-redis object as a "default", then we'd have to maintain 2 additional shards (i.e. `cable-cr/jgaskins-redis-adapter`, `cable-cr/stefanwille-redis-adapter`) Doable for sure, but also a bit of a pain to manage (for me).

I agree with the type handling. There's this issue https://github.com/jgaskins/redis/issues/32 which I think will make that a lot nicer. 

I appreciate the info @mjeffrey18! It sounds like you're running in to a lot of the same stuff we are, so that's a huge help :pray:

File: src/cable/handler.cr [hunk 3]
```diff
@@ -36,17 +44,28 @@ module Cable
...
+            # handle unknown/malformed messages
+            socket.close(HTTP::WebSocket::CloseCode::InvalidFramePayloadData, "Invalid message")
+            Cable::Logger.error { "KeyError Exception: #{e.message}" }
           rescue e : Cable::Connection::UnathorizedConnectionException
             # do nothing, this is planned
+            socket.close(HTTP::WebSocket::CloseCode::NormalClosure, "Farewell")
+          rescue e : IO::Error
+            Cable::Logger.error { "#{e.class.name} Exception: #{e.message} -> #{self.class.name}#call { socket.on_message(message) }" }
+            # Redis may have some error, restart Cable
+            socket.close(HTTP::WebSocket::CloseCode::NormalClosure, "Farewell")
```
Comment (jwoertink) [inline src/cable/handler.cr:57]: In #54 @mjeffrey18 added the `socket.close`, but this rescue branch doesn't exist on master.... Should we be closing the socket here too? If so, then which CloseCode should be used?

**Comment by jwoertink** [comment]
This PR is no longer necessary. The current implementation allows for choosing which Redis shard you want to use.

**Pull request closed without merging**